json_rpc/src/request.rs

```diff
@@ -0,0 +1,208 @@
+use std::convert::TryFrom;
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use warp::reject::{self, Rejection};
+
+use crate::{
+    error::{Error, ReservedErrorCode},
+    rejections::MissingId,
+    JSON_RPC_VERSION,
+};
+
+/// A JSON-RPC request, prior to validation of conformance to the JSON-RPC specification.
+///
+/// This overly-permissive type used in the warp filters allows for non-conformance to be detected
+/// and a useful error message returned to the client, rather than a somewhat opaque HTTP 404
+/// response.
+#[derive(Eq, PartialEq, Serialize, Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub(crate) struct UnvalidatedRequest {
+    jsonrpc: String,
+    id: Option<Value>,
+    method: String,
+    params: Option<Value>,
+}
+
+/// Errors are returned to the client as a JSON-RPC response and HTTP 200 (OK), whereas rejections
+/// cause no JSON-RPC response to be sent, but an appropriate HTTP 4xx error will be returned.
+#[derive(Debug)]
+pub(crate) enum ErrorOrRejection {
+    Error { id: Value, error: Error },
+    Rejection(Rejection),
+}
+
+/// A request which has been validated as conforming to the JSON-RPC specification.
+pub(crate) struct Request {
+    pub id: Value,
+    pub method: String,
+    pub params: Option<Value>,
+}
+
+/// Returns `Ok` if `id` is a String, Null or a Number with no fractional part.
+fn is_valid(id: &Value) -> Result<(), Error> {
```

Comment:
maybe we can put that in its own `mod id` module? Its API looks like it can be used on any `Value` type.

Reply:
That's actually how `warp_json_rpc` handles it (i.e. it provides an `Id` enum) but the downside is that parsing from an incoming JSON-RPC request where the "id" field is invalid in the warp filter will fail if we use a strict enum.  This causes a warp rejection, which _can_ be handled later in the flow, but the useful info about why parsing failed is lost.

I prefer this more permissive parsing, so we can handle failures by returning useful data to the user in the error of the JSON-RPC response.